File: 42-trapping-rain-water/42-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        if n <=2:
            return 0
        left = [0]*n
        right = [0]*n
        left[0] = 0
        leftMax = height[0]
        for i in range(n):
            left[i] = leftMax
            leftMax = max(leftMax,height[i])
        right[n-1] = 0
        rightMax = height[n-1]
        for i in range(n-2,-1,-1):
            right[i] = rightMax
            rightMax = max(rightMax,height[i])
        
        trappedWater = 0
        for i in range(n):
            if height[i] < left[i] and height[i] < right[i]:
                trappedWater +=min(left[i],right[i]) - height[i]
        return trappedWater
                

File: 42-trapping-rain-water/42-trapping-rain-water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        if(n<=2)
            return 0;
        int maxLeft = height[0];
        int maxRight = height[n-1];
        int trappedWater = 0;
        int left = 1;
        int right = n-2;
        while (left <= right){
            if (maxLeft < maxRight){
                if (height[left] >= maxLeft)
                    maxLeft = height[left];
                
                else{
                    trappedWater +=maxLeft - height[left];}
                left +=1;
                }
            else{
                if (height[right] >=maxRight)
                    maxRight = height[right];
                else
                    trappedWater += maxRight - height[right];
                right -= 1;
                }
            }
        return trappedWater;
    }
};